### DISCUSSION TITLE
Add ability to OK/cancel Color Picker with a sensible way

### DISCUSSION THREAD

**Discussion by miv391** [godotengine/godot-proposals#8084] [Editor]
### Problem

Currently when you open the Color Picker from the Inspector, you have following options:

1. Press Esc to accept the new color 
2. Click outside of the Color Picker to accept the new color 

![kuva](https://github.com/godotengine/godot-proposals/assets/49998025/dd62486a-6621-4d4f-a982-97ce6159e616)

Problems:
- The only options are to accept the new color. If you happened to change the color in the Color Picker but don't want to use it after all, you still need to accept it and then press Ctrl-Z _twice_ to undo the color change.
- The only way to close the Color Picker with a mouse is to click somewhere outside of it. At least to me it always feels awkward and uncomfortable to click "somewhere". At least you cannot break anything with that click.
- The shortcut key used for accepting the new color is Esc. Probably the worst possible option.

How other similar dialogs (although Color Picker is a popup) work: If you open the Script / Quick Load dialog (titled Resource) from the Inspector, it has these options:

1. Select script from list and set it to the property's new value
  - by clicking Open button
  - or by pressing Enter
2. Cancel the whole operation without changing anything
  - by clicking Cancel button
  - or by pressing Esc

### Proposal

I propose that the Inspector's Color Picker usability is fixed by changing it to work like the Script / Quick Load dialog.

- OK and Cancel buttons are added to the bottom of Color Picker. Both buttons close the Color Picker.
	- If OK button is clicked, color property is changed to the selected color.
	- If Cancel button is clicked, color property is not changed .
- Enter is the shortcut key for the OK button.
- Esc is the shortcut key for the Cancel button. 

I don't think that the ColorPicker node itself should contain OK/Cancel buttons. This proposal is about how the Inspector's color picking works. I'm not familiar with the editor code, but that _may_ mean the ColorPickerButton node.

I modified this addon https://github.com/Qubus0/GodotColorPreview/tree/main to work like I proposed. I added the shortcut keys to the button texts, because this color picker now works differently than the Inspector Color Picker. This is much more pleasant to use than the Inspector's Color Picker.

![kuva](https://github.com/godotengine/godot-proposals/assets/49998025/f9a21091-5996-4d64-a266-44a9f8ed33b2)

**Comment by aXu-AP** [comment]
I think the distinction which you made, that this is a popup and not a dialog, is important. I expect modifications made in the main window (and popups are extension of that window) to take effect immediately ie. as it currently is. Having used Photoshop, I always felt that choosing colors in a dialog and then pressing enter/accept is sluggish.

> then press Ctrl-Z twice to undo the color change.

This sounds like a bug to me.

**Reply by miv391** [reply]
That is a valid point. Still, the usability of current color picker is bad. Dropdown lists are quite similar as this popup, but they work like this:
1. set focus to some item (just by moving mouse over the item)
2. press Enter to select the focused item and close the dropdown OR press Esc to close the dropdown without changing selection

In my opinion, the logic is different. Color picker could work in similar way:
- click color picker button OR press Enter to confirm the new selected color
- click somewhere else than the color picker button OR press Esc to cancel the color selection

Maybe users are too accustomed to close the color picker by clicking outside of the color picker that it should also be considered as selecting a new color, but I would like to have at least one way to cancel the color selection.

**Comment by Calinou** [comment]
Adding OK/Cancel buttons makes this already large popup even larger, so I'd rather not have that.

> - The only options are to accept the new color. If you happened to change the color in the Color Picker but don't want to use it after all, you still need to accept it and then press Ctrl-Z twice to undo the color change.

Like in Photoshop, you can click the "old" color on the left of the preview in the middle to revert to the old color without closing the color picker. This could benefit from a tooltip or other visible indication (such as a revert icon) though.

> - The only way to close the Color Picker with a mouse is to click somewhere outside of it. At least to me it always feels awkward and uncomfortable to click "somewhere". At least you cannot break anything with that click.

I guess we could make <kbd>Ctrl + Enter</kbd> accept the current suggestion in any of the text fields and close the color picker in a manner similar to <kbd>Escape</kbd>. <kbd>Ctrl + Enter</kbd> is generally understood as a "submit" interaction (try it on GitHub :slightly_smiling_face:).

**Reply by miv391** [reply]
I hadn't noticed that old color swatch al all. That at least makes it possible to revert to the original color.

**Reply by Calinou** [reply]
Pull requests opened:

- https://github.com/godotengine/godot/pull/84447
- https://github.com/godotengine/godot/pull/84527
